crates/libs/registry/src/transaction.rs

```diff
@@ -0,0 +1,53 @@
+use super::*;
+
+/// A transaction object.
+#[repr(transparent)]
+#[derive(Debug)]
+pub struct Transaction(pub(crate) HANDLE);
+
+impl Transaction {
+    /// Creates a new transaction.
+    pub fn new() -> Result<Self> {
+        let handle = unsafe { CreateTransaction(null_mut(), null_mut(), 0, 0, 0, 0, null()) };
+
+        if handle == INVALID_HANDLE_VALUE {
+            Err(Error::from_win32())
+        } else {
+            Ok(Self(handle))
+        }
+    }
+
+    /// Commits the transaction.
+    ///
+    /// The transaction rolls back if it is dropped before `commit` is called.
+    pub fn commit(self) -> Result<()> {
+        let result = unsafe { CommitTransaction(self.0) };
+
+        if result == 0 {
+            Err(Error::from_win32())
+        } else {
+            Ok(())
+        }
+    }
```

Comment:
Thoughts on automatically rolling back the transaction on failure?

Reply:
Drop the transaction object and it rolls back automatically.

Reply:
After `CommitTransaction` fails, the transaction is in a limbo unknown state, and there's no way to retrieve that state. So my thinking was to just explicitly rollback all failed commits to make the state known. (Granted, developers would lose the ability to retry the transaction.)

Reply:
Recovery is typically impractical thus simply unwind and drop everything, starting again if desired.